### PR TITLE
Added default value support

### LINKED ISSUE #87: Case class with Seq parameter default = Nil and  not supported, deserialized to null
# Given case class:

case class Cat(name : String, kittens: Seq[Cat] = Nil)
# Read in test:

mapper.readValue("{\"name\":\"kozmo\"}", classOf[Cat])
# Expected:

expected cat with Nil kittens, however it is null. it seems like SeqModule does not check for the default value if the parameter does not provieded

### PR DESCRIPTION
**Pull request by jroper**
Fixes #87

### CODE REVIEW

**Comment by jroper** [comment]
The build failure occurs on the 2.11 branch without this change. Should this change be made to the 2.11 branch? I didn't make it on the master branch because that doesn't even compile. Would the 2.10 branch be better?

**Comment by jroper** [comment]
One thing about this change, if you have:

```scala
case class Foo(bar: Option[String] = Some("bar"))
```

And you don't pass `bar`, you'll get `Some("bar")`, not `None`. I could make it so you get `None`, but not sure if that should be the desired behavior or not.

**Comment by pjfanning** [comment]
master branch is for the 3.0 release and it needs plenty of work to get all the tests to pass.

2.11 branch is the best place to start. The build was passing last month but it does seem to be broken now. It might be a few days before I can investigate the issue.

**Comment by pjfanning** [comment]
@cowtowncoder do you know of any changes in jackson-databind 2.11.0-SNAPSHOT that could be causing this? The tests were passing a month ago and no jackson-module-scala have been done.

```
[info] - should be able to deserialize left with null value *** FAILED ***
[info]   com.fasterxml.jackson.databind.exc.InvalidDefinitionException: Cannot find a Value deserializer for abstract type [reference type, class scala.util.Either<scala.util.Either<java.lang.String,java.lang.String>><[simple type, class scala.util.Either<java.lang.String,java.lang.String>]>]
[info]  at [Source: (String)"["left", null]"; line: 1, column: 1]
[info]   at com.fasterxml.jackson.databind.exc.InvalidDefinitionException.from(InvalidDefinitionException.java:67)
[info]   at com.fasterxml.jackson.databind.DeserializationContext.reportBadDefinition(DeserializationContext.java:1596)
[info]   at com.fasterxml.jackson.databind.deser.DeserializerCache._handleUnknownValueDeserializer(DeserializerCache.java:591)
[info]   at com.fasterxml.jackson.databind.deser.DeserializerCache.findValueDeserializer(DeserializerCache.java:148)
[info]   at com.fasterxml.jackson.databind.DeserializationContext.findRootValueDeserializer(DeserializationContext.java:476)
[info]   at com.fasterxml.jackson.databind.ObjectMapper._findRootDeserializer(ObjectMapper.java:4389)
[info]   at com.fasterxml.jackson.databind.ObjectMapper._readMapAndClose(ObjectMapper.java:4198)
[info]   at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:3205)
[info]   at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:3188)
[info]   at com.fasterxml.jackson.module.scala.deser.DeserializerTest.deserialize(DeserializerTest.scala:13)
```

File: src/main/scala/com/fasterxml/jackson/module/scala/introspect/ScalaAnnotationIntrospectorModule.scala [hunk 1]
```diff
@@ -114,9 +117,54 @@ object ScalaAnnotationIntrospector extends NopAnnotationIntrospector
...
+
+    if (isMaybeScalaBeanType(beanDesc.getBeanClass)) {
+
+      val descriptor = _descriptorFor(beanDesc.getBeanClass)
+      if (descriptor.properties.exists(_.param.exists(_.defaultValue.isDefined))) {
+        defaultInstantiator match {
+          case std: StdValueInstantiator =>
+            new ScalaValueInstantiator(std, config, descriptor)
+          case other =>
+            throw new IllegalArgumentException("Cannot customise a non StdValueInstantiatiator: " + other.getClass)
```
Comment (jroper) [inline src/main/scala/com/fasterxml/jackson/module/scala/introspect/ScalaAnnotationIntrospectorModule.scala]: The Kotlin module does exactly the same thing here, and it's been doing that for 4 years presumably with no problems, so I think we're fine to fail too:

https://github.com/FasterXML/jackson-module-kotlin/blob/3eb66acf2638dfd99a53a3d6e1389844b5caed2e/src/main/kotlin/com/fasterxml/jackson/module/kotlin/KotlinValueInstantiator.kt#L174

**Comment by cowtowncoder** [comment]
@pjfanning Can not think of anything specific that would trigger such a failure, offhand. :-(

**Comment by pjfanning** [comment]
the existing jackson-module-scala 2.11 branch tests pass when run against jackson-databind 2.10.1

**Comment by jroper** [comment]
Thanks! Should I submit another PR to forward port to master? And is there any chance this could be backported to 2.10.x or will we wait for 2.11?

**Comment by pjfanning** [comment]
@jroper this does affect existing users, who might have case classes with default values but rely on the fact that jackson does not support this - could we make the behaviour configurable? If it was configurable, then it could disabled by default in the 2.10 backport and enabled by default in 2.11.

**Comment by pjfanning** [comment]
@jroper when ported to master branch, the unit test has never seemed to work - the master branch is for the yet-to-be-released jackson v3. Your code works well in jackson v2. If you have time would you be able to look at the issue in master? The test is currently commented out. The issue could be a mistake I made when porting the code but I can't see anything obvious.

**Pull request merged**